Sources/Basic/OrderedSet.swift

```diff
@@ -0,0 +1,102 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+/// An ordered set is an ordered collection of instances of `Element` in which
+/// uniqueness of the objects is guaranteed.
+public struct OrderedSet<E: Hashable>: Equatable, Collection {
+    public typealias Element = E
+    public typealias Index = Int
+    public typealias Indices = CountableRange<Int>
+
+    private var array: [Element]
+    private var set: Set<Element>
+    
+    /// Creates an empty ordered set.
+    public init() {
+        self.array = []
+        self.set = Set()
+    }
+    
+    /// Creates an ordered set with the contents of `array`.
+    ///
+    /// If an element occurs more than once in `element`, only the first one
+    /// will be included.
+    public init(_ array: [Element]) {
+        self.init()
+        for element in array {
+            append(element)
+        }
+    }
+    
+    // MARK: Working with an ordered set
+    
+    /// The number of elements the ordered set stores.
+    public var count: Int { return array.count }
+    
+    /// Returns `true` if the set is empty.
+    public var isEmpty: Bool { return array.isEmpty }
+    
+    /// Returns the contents of the set as an array.
+    public var contents: [Element] { return array }
+    
+    /// Returns `true` if the ordered set contains `member`.
+    public func contains(_ member: Element) -> Bool {
+        return set.contains(member)
+    }
+    
+    /// Adds an element to the ordered set.
+    ///
+    /// If it already contains the element, then the set is unchanged.
+    ///
+    /// - returns: True if the item was inserted.
+    @discardableResult
+    public mutating func append(_ newElement: Element) -> Bool {
+        let inserted = set.insert(newElement).inserted
+        if inserted {
+            array.append(newElement)
+        }
+        return inserted
+    }
+    
+    /// Remove and return the element at the end of the ordered set.
+    public mutating func removeLast() -> Element {
+        let lastElement = array.removeLast()
+        set.remove(lastElement)
+        return lastElement
+    }
+    
+    /// Remove all elements.
+    public mutating func removeAll(keepingCapacity capacity: Bool) {
```

Comment:
Nit: "capacity" is an odd name of a boolean here. Usually in a collection, "capacity" indicates how many slots (or whatever is appropriate) to allocate.  "keep" would be better.  Again, just a nit but this would be the time to fix it...